From 7b967cdbc86aa033088ebb39ad366388c17447de Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Horia=20Geant=C4=83?= <horia.geanta@nxp.com>
Date: Thu, 19 Apr 2018 16:47:48 +0300
Subject: [PATCH 563/706] crypto: caam/qi2 - fix non-hmac hashes
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Since 6de62f15b581 ("crypto: algif_hash - Require setkey before
accept(2)"), the AF_ALG interface requires userspace to provide a key
to any algorithm that has a setkey method.
More, since 9fa68f620041 ("crypto: hash - prevent using keyed hashes
without setting key") crypto API started behaving in a similar way.

However, the non-HMAC algorithms are not keyed, so setting a key is
unnecessary.
Fix this by removing the setkey method from the non-keyed hash
algorithms.

Fixes: 254c9ec0b628 ("crypto: caam/qi2 - add support for ahash algorithms")
Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
(cherry picked from commit d8c73a1dc560f1ae3408163b14754175c7fd6a19)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/crypto/caam/caamalg_qi2.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/crypto/caam/caamalg_qi2.c b/drivers/crypto/caam/caamalg_qi2.c
index ae5eafdb890b..e3f5fc8bc765 100644
--- a/drivers/crypto/caam/caamalg_qi2.c
+++ b/drivers/crypto/caam/caamalg_qi2.c
@@ -5083,6 +5083,7 @@ static struct caam_hash_alg *caam_hash_alloc(struct device *dev,
 			 template->name);
 		snprintf(alg->cra_driver_name, CRYPTO_MAX_ALG_NAME, "%s",
 			 template->driver_name);
+		t_alg->ahash_alg.setkey = NULL;
 	}
 	alg->cra_module = THIS_MODULE;
 	alg->cra_init = caam_hash_cra_init;
-- 
2.17.1

